### PR TITLE
Optimizations for better promotion to affine dialect

### PR DESCRIPTION
**Pull request by rjnw**
Two separate optimization pass:
 * Optimizing removing unused loop iteration variables
 * load and store forwarding

Three rewrite patterns for convert operations.

updating affine promotion for shape and shape shift operations.

### CODE REVIEW

File: flang/include/flang/Optimizer/Transforms/Passes.td [hunk 1]
```diff
@@ -43,6 +43,20 @@ def AffineDialectDemotion : FunctionPass<"demote-affine"> {
   let constructor = "fir::createAffineDemotionPass()";
 }
 
+def FirLoopResultOpt : FunctionPass<"fir-loop-result-opt"> {
+    let constructor = "fir::createFirLoopResultOptPass()";
```
Comment (schweitzpgi) [inline flang/include/flang/Optimizer/Transforms/Passes.td:47]: Maybe add a summary and description

File: flang/lib/Optimizer/Transforms/AffinePromotion.cpp [hunk 2]
```diff
@@ -504,6 +533,9 @@ class AffineDialectPromotion
 
     auto *context = &getContext();
     auto function = getFunction();
+    markAllAnalysesPreserved();
+    Liveness l = Liveness(function);
+    // l.dump();
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Transforms/AffinePromotion.cpp:538]: LLVM_DEBUG() ?

File: flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    rewriter.startRootUpdate(loop.getOperation());
+    newLoop.getBody()->getOperations().splice(
+        --newLoop.getBody()->end(), loopOps, loopOps.begin(), --loopOps.end());
+    loop.getInductionVar().replaceAllUsesWith(newLoop.getInductionVar());
+    rewriter.finalizeRootUpdate(loop.getOperation());
+    rewriter.finalizeRootUpdate(newLoop.getOperation());
+    for (auto r : loop.getResults()) {
+      eraseAllUses(r, rewriter);
+    }
+    rewriter.eraseBlock(loop.getBody());
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp:49]: Does this do a dropAllReferences() on the Region for you?

File: flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    auto *context = &getContext();
+    auto function = getFunction();
+    mlir::OwningRewritePatternList patterns;
+    patterns.insert<LoopResultRemoval>(context);
+    mlir::ConversionTarget target = *context;
+    target.addLegalDialect<fir::FIROpsDialect, mlir::StandardOpsDialect>();
+    target.addDynamicallyLegalOp<fir::DoLoopOp>([&](fir::DoLoopOp op) {
+      if (op.getNumResults() == 0)
+        return true;
+      return false;
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp:94]: Could be a 1-line `return`.

File: flang/include/flang/Optimizer/Transforms/Passes.td (hunk 1 above)
Comment (rjnw) [reply flang/include/flang/Optimizer/Transforms/Passes.td:47]: okay, also I am not sure about the name of this pass.

File: flang/lib/Optimizer/Transforms/AffinePromotion.cpp (hunk 2 above)
Comment (rjnw) [reply flang/lib/Optimizer/Transforms/AffinePromotion.cpp:538]: I just added this here when I was playing around with liveness. I will remove it.

File: flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp (hunk 3 above)
Comment (rjnw) [reply flang/lib/Optimizer/Transforms/FirLoopResultOpt.cpp:49]: yes, `dropAllReferences` looks better. Thank you!

File: flang/include/flang/Optimizer/Transforms/Passes.td (hunk 1 above)
Comment (schweitzpgi) [reply flang/include/flang/Optimizer/Transforms/Passes.td:47]: Ah. Maybe something like "DCE loop results"? That may be worse though. heh

**Pull request merged**